File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonElectionSerializer.java
package com.github.dedis.popstellar.model.network.serializer;

import com.github.dedis.popstellar.model.network.method.message.data.election.*;
import com.github.dedis.popstellar.model.objects.Channel;
import com.github.dedis.popstellar.model.objects.Election;
import com.github.dedis.popstellar.model.objects.event.EventState;
import com.github.dedis.popstellar.model.objects.security.MessageID;
import com.github.dedis.popstellar.model.objects.security.PublicKey;
import com.google.gson.*;

import java.lang.reflect.Type;
import java.util.*;

public class JsonElectionSerializer
    implements JsonSerializer<Election>, JsonDeserializer<Election> {

  @Override
  public Election deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
      throws JsonParseException {
    JsonObject jsonObject = json.getAsJsonObject();

    // Deserialize the nested Channel object
    Channel channel = context.deserialize(jsonObject.get("channel"), Channel.class);

    // Deserialize the primitive values
    String id = jsonObject.get("id").getAsString();
    String name = jsonObject.get("name").getAsString();
    long creation = jsonObject.get("creation").getAsLong();
    long start = jsonObject.get("start").getAsLong();
    long end = jsonObject.get("end").getAsLong();
    JsonElement electionKeyElement = jsonObject.get("electionKey");
    String electionKey = electionKeyElement == null ? null : electionKeyElement.getAsString();

    // Deserialize the enums
    ElectionVersion electionVersion =
        ElectionVersion.valueOf(jsonObject.get("electionVersion").getAsString());
    String stateString = jsonObject.get("state").getAsString();
    EventState state = stateString.isEmpty() ? null : EventState.valueOf(stateString);

    // Deserialize collections
    JsonArray electionQuestionsJsonArray = jsonObject.get("electionQuestions").getAsJsonArray();
    List<ElectionQuestion> electionQuestions = new ArrayList<>();
    for (JsonElement electionQuestionJsonElement : electionQuestionsJsonArray) {
      electionQuestions.add(
          context.deserialize(electionQuestionJsonElement, ElectionQuestion.class));
    }

    Map<PublicKey, List<Vote>> votesBySender = new HashMap<>();
    JsonObject votesBySenderObject = jsonObject.get("votesBySender").getAsJsonObject();
    for (Map.Entry<String, JsonElement> entry : votesBySenderObject.entrySet()) {
      PublicKey senderPublicKey = new PublicKey(entry.getKey());
      JsonArray votesBySenderJsonArray = entry.getValue().getAsJsonArray();
      List<Vote> votesBySenderPk = new ArrayList<>();
      for (JsonElement element : votesBySenderJsonArray) {
        votesBySenderPk.add(context.deserialize(element, Vote.class));
      }
      votesBySender.put(senderPublicKey, votesBySenderPk);
    }

    Map<PublicKey, MessageID> messageMap = new HashMap<>();
    JsonObject messageMapObject = jsonObject.get("messageMap").getAsJsonObject();
    for (Map.Entry<String, JsonElement> entry : messageMapObject.entrySet()) {
      PublicKey senderPublicKey = new PublicKey(entry.getKey());
      MessageID messageId = new MessageID(entry.getValue().getAsString());
      messageMap.put(senderPublicKey, messageId);
    }

    Map<String, Set<QuestionResult>> results = new HashMap<>();
    JsonObject resultsObject = jsonObject.get("results").getAsJsonObject();
    for (Map.Entry<String, JsonElement> entry : resultsObject.entrySet()) {
      String questionId = entry.getKey();
      JsonArray resultsJsonArray = entry.getValue().getAsJsonArray();
      Set<QuestionResult> resultsByQuestion = new HashSet<>();
      for (JsonElement element : resultsJsonArray) {
        resultsByQuestion.add(context.deserialize(element, QuestionResult.class));
      }
      results.put(questionId, resultsByQuestion);
    }

    return new Election(
        id,
        name,
        creation,
        channel,
        start,
        end,
        electionQuestions,
        electionKey,
        electionVersion,
        votesBySender,
        messageMap,
        state,
        results);
  }

  @Override
  public JsonElement serialize(
      Election election, Type typeOfSrc, JsonSerializationContext context) {
    JsonObject jsonObject = new JsonObject();

    jsonObject.add("channel", context.serialize(election.getChannel(), Channel.class));

    jsonObject.addProperty("id", election.getId());
    jsonObject.addProperty("name", election.getName());
    jsonObject.addProperty("creation", election.getCreation());
    jsonObject.addProperty("start", election.getStartTimestamp());
    jsonObject.addProperty("end", election.getEndTimestamp());
    jsonObject.addProperty("electionKey", election.getElectionKey());

    EventState state = election.getState();
    jsonObject.addProperty("state", state == null ? "" : state.name());
    jsonObject.addProperty("electionVersion", election.getElectionVersion().name());

    JsonArray electionQuestionsJsonArray = new JsonArray();
    for (ElectionQuestion electionQuestion : election.getElectionQuestions()) {
      electionQuestionsJsonArray.add(context.serialize(electionQuestion, ElectionQuestion.class));
    }
    jsonObject.add("electionQuestions", electionQuestionsJsonArray);

    JsonObject votesBySenderJsonObject = new JsonObject();
    for (Map.Entry<PublicKey, List<Vote>> entry : election.getVotesBySender().entrySet()) {
      String pk = entry.getKey().getEncoded();
      JsonArray votesBySenderJsonArray = new JsonArray();
      for (Vote vote : entry.getValue()) {
        votesBySenderJsonArray.add(context.serialize(vote, Vote.class));
      }
      votesBySenderJsonObject.add(pk, votesBySenderJsonArray);
    }
    jsonObject.add("votesBySender", votesBySenderJsonObject);

    JsonObject messageMapJsonObject = new JsonObject();
    for (Map.Entry<PublicKey, MessageID> entry : election.getMessageMap().entrySet()) {
      String pk = entry.getKey().getEncoded();
      String messageID = entry.getValue().getEncoded();
      JsonElement jsonElement = new JsonPrimitive(messageID);
      messageMapJsonObject.add(pk, jsonElement);
    }
    jsonObject.add("messageMap", messageMapJsonObject);

    JsonObject resultsJsonObject = new JsonObject();
    for (Map.Entry<String, Set<QuestionResult>> entry : election.getResults().entrySet()) {
      JsonArray resultsJsonArray = new JsonArray();
      for (QuestionResult vote : entry.getValue()) {
        resultsJsonArray.add(context.serialize(vote, QuestionResult.class));
      }
      resultsJsonObject.add(entry.getKey(), resultsJsonArray);
    }
    jsonObject.add("results", resultsJsonObject);

    return jsonObject;
  }
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/database/JsonElectionSerializer.kt
package com.github.dedis.popstellar.model.network.serializer.database

import com.github.dedis.popstellar.model.network.method.message.data.election.ElectionQuestion
import com.github.dedis.popstellar.model.network.method.message.data.election.ElectionVersion
import com.github.dedis.popstellar.model.network.method.message.data.election.QuestionResult
import com.github.dedis.popstellar.model.network.method.message.data.election.Vote
import com.github.dedis.popstellar.model.objects.Channel
import com.github.dedis.popstellar.model.objects.Election
import com.github.dedis.popstellar.model.objects.event.EventState
import com.github.dedis.popstellar.model.objects.security.MessageID
import com.github.dedis.popstellar.model.objects.security.PublicKey
import com.google.gson.JsonArray
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParseException
import com.google.gson.JsonPrimitive
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import java.lang.reflect.Type

class JsonElectionSerializer : JsonSerializer<Election>, JsonDeserializer<Election> {
  @Throws(JsonParseException::class)
  override fun deserialize(
      json: JsonElement,
      typeOfT: Type,
      context: JsonDeserializationContext
  ): Election {
    val jsonObject = json.asJsonObject

    // Deserialize the nested Channel object
    val channel = context.deserialize<Channel>(jsonObject["channel"], Channel::class.java)

    // Deserialize the primitive values
    val id = jsonObject["id"].asString
    val name = jsonObject["name"].asString
    val creation = jsonObject["creation"].asLong
    val start = jsonObject["start"].asLong
    val end = jsonObject["end"].asLong
    val electionKeyElement = jsonObject["electionKey"]
    val electionKey = electionKeyElement?.asString

    // Deserialize the enums
    val electionVersion = ElectionVersion.valueOf(jsonObject["electionVersion"].asString)
    val stateString = jsonObject["state"].asString
    val state = if (stateString.isEmpty()) null else EventState.valueOf(stateString)

    // Deserialize the list electionQuestions
    val electionQuestionsJsonArray = jsonObject["electionQuestions"].asJsonArray
    val electionQuestions: MutableList<ElectionQuestion> = ArrayList()
    for (electionQuestionJsonElement in electionQuestionsJsonArray) {
      electionQuestions.add(
          context.deserialize(electionQuestionJsonElement, ElectionQuestion::class.java))
    }

    // Deserialize the map votesBySender
    val votesBySender: MutableMap<PublicKey, List<Vote>> = HashMap()
    val votesBySenderObject = jsonObject["votesBySender"].asJsonObject
    for ((key, value) in votesBySenderObject.entrySet()) {
      val senderPublicKey = PublicKey(key)
      val votesBySenderJsonArray = value.asJsonArray
      // Deserialize the value of the map (list of votes)
      val votesBySenderPk: MutableList<Vote> = ArrayList()
      for (element in votesBySenderJsonArray) {
        votesBySenderPk.add(context.deserialize(element, Vote::class.java))
      }
      votesBySender[senderPublicKey] = votesBySenderPk
    }

    // Deserialize the map messageMap
    val messageMap: MutableMap<PublicKey, MessageID> = HashMap()
    val messageMapObject = jsonObject["messageMap"].asJsonObject
    for ((key, value) in messageMapObject.entrySet()) {
      val senderPublicKey = PublicKey(key)
      val messageId = MessageID(value.asString)
      messageMap[senderPublicKey] = messageId
    }

    // Deserialize the map results
    val results: MutableMap<String, Set<QuestionResult>> = HashMap()
    val resultsObject = jsonObject["results"].asJsonObject
    for ((questionId, value) in resultsObject.entrySet()) {
      // Deserialize the value of the map (set of question results)
      val resultsJsonArray = value.asJsonArray
      val resultsByQuestion: MutableSet<QuestionResult> = HashSet()
      for (element in resultsJsonArray) {
        resultsByQuestion.add(context.deserialize(element, QuestionResult::class.java))
      }
      results[questionId] = resultsByQuestion
    }

    return Election(
        id,
        name,
        creation,
        channel,
        start,
        end,
        electionQuestions,
        electionKey,
        electionVersion,
        votesBySender,
        messageMap,
        state,
        results)
  }

  override fun serialize(
      election: Election,
      typeOfSrc: Type,
      context: JsonSerializationContext
  ): JsonElement {
    val jsonObject = JsonObject()

    // Serialize the channel object
    jsonObject.add("channel", context.serialize(election.channel, Channel::class.java))

    // Serialize the primitive values
    jsonObject.addProperty("id", election.id)
    jsonObject.addProperty("name", election.name)
    jsonObject.addProperty("creation", election.creation)
    jsonObject.addProperty("start", election.startTimestamp)
    jsonObject.addProperty("end", election.endTimestamp)
    jsonObject.addProperty("electionKey", election.electionKey)

    // Serialize the enum
    val state = election.state
    jsonObject.addProperty("state", state?.name ?: "")
    jsonObject.addProperty("electionVersion", election.electionVersion.name)

    // Serialize the list of election questions into a JsonArray
    val electionQuestionsJsonArray = JsonArray()
    for (electionQuestion in election.electionQuestions) {
      electionQuestionsJsonArray.add(
          context.serialize(electionQuestion, ElectionQuestion::class.java))
    }
    jsonObject.add("electionQuestions", electionQuestionsJsonArray)

    // Serialize the map of votesBySender into a JsonObject
    val votesBySenderJsonObject = JsonObject()
    for ((key, value) in election.votesBySender) {
      val pk = key.encoded
      // Serialize the list of votes into a JsonArray
      val votesBySenderJsonArray = JsonArray()
      for (vote in value) {
        votesBySenderJsonArray.add(context.serialize(vote, Vote::class.java))
      }
      votesBySenderJsonObject.add(pk, votesBySenderJsonArray)
    }
    jsonObject.add("votesBySender", votesBySenderJsonObject)

    // Serialize the messageMap into a JsonObject
    val messageMapJsonObject = JsonObject()
    for ((key, value) in election.messageMap) {
      val pk = key.encoded
      val messageID = value.encoded
      val jsonElement: JsonElement = JsonPrimitive(messageID)
      messageMapJsonObject.add(pk, jsonElement)
    }
    jsonObject.add("messageMap", messageMapJsonObject)

    // Serialize the results map into a JsonObject
    val resultsJsonObject = JsonObject()
    for ((key, value) in election.results) {
      // Serialize the set of question results into a JsonArray
      val resultsJsonArray = JsonArray()
      for (vote in value) {
        resultsJsonArray.add(context.serialize(vote, QuestionResult::class.java))
      }
      resultsJsonObject.add(key, resultsJsonArray)
    }
    jsonObject.add("results", resultsJsonObject)

    return jsonObject
  }
}
